File: php-poo/aulas/aula15/Visualizacao.php
<?php
require_once 'Video.php';
require_once 'Gafanhoto.php';
class Visualizacao
{
    // Attribute
    private $espectador;
    private $filme;

    // Method Construct
    public function __construct($espectador, $filme)
    {
        $this->espectador = $espectador;
        $this->filme = $filme;
        $this->filme->setViews($this->filme->getViews() + 1); // contar o $views (Classe Video)
        $this->espectador->setTotalAssistido($this->espectador->getTotalAssistido() + 1 ); // contar o $totalAssistido (Classe Gaganhoto)
    }

    // @override @Sobrescrever
    public function avaliar() {
        $this->filme->setAvaliacao(5);
    }

    public function avaliarNota($nota) {
        $this->filme->setAvaliacao($nota);
    }

    public function avaliarPorc($porc) {
        $nova = 0;
        if ($porc <= 20) {
            $nova = 3;
        } elseif ($porc <= 50) {
            $nova = 5;
        } elseif ($porc <= 90) {
            $nova = 8;
        } else {
            $nova = 10;
        }
        $this->filme->setAvaliacao($nova);
    }


    // Special Methods, Getters and Setters
    public function getEspectador()
    {
        return $this->espectador;
    }

    public function setEspectador($espectador): void
    {
        $this->espectador = $espectador;
    }

    public function getFilme()
    {
        return $this->filme;
    }

    public function setFilme($filme): void
    {
        $this->filme = $filme;
    }

}

File: java-poo/projects/poo-aula15/src/aula15/Visualizacao.java
package aula15;

public class Visualizacao {
    // Agregação
    private Gafanhoto espectador;
    private Video filme;

    // Constructor
    public Visualizacao(Gafanhoto espectador, Video filme) {
        this.espectador = espectador;
        this.filme = filme;
        this.espectador.setTotalAssistido(this.espectador.getTotalAssistido() + 1);
        this.filme.setViews(this.filme.getViews() + 1);
    }

    // Methods
    // @SobreCarga - 3 metodos com mesmo nome e assinaturas diferentes na mesma classe.
    public void avaliar() {
        this.filme.setAvaliacao(5);
    }
    public void avaliar(int nota) {
        this.filme.setAvaliacao(nota);
    }
    public void avaliar(float porcentagem) {
        int total = 0;
        if ( porcentagem <= 20 ) total = 3;
        else if ( porcentagem <= 30 ) total = 5;
        else if ( porcentagem <= 90 ) total = 8;
        else total = 10;

        this.filme.setAvaliacao(total);
    }

    // Special Methods - Getters and Setters
    public Gafanhoto getEspectador() {
        return espectador;
    }

    public void setEspectador(Gafanhoto espectador) {
        this.espectador = espectador;
    }

    public Video getFilme() {
        return filme;
    }

    public void setFilme(Video filme) {
        this.filme = filme;
    }

    @Override
    public String toString() {
        return "Visualizacao{" +
                "espectador=" + espectador +
                ", filme=" + filme +
                '}';
    }
}
